### ISSUE TITLE
[question|help] How to use this image while using `network_mode: service:another_service`?

### ISSUE THREAD

**Issue by zhyph** [joesturge/lazymc-docker-proxy#94]
Hello,

I have a "somewhat" unique setup where I use an existent wireguard container to share my minecraft server with some friends, all the port forwarding and connections are done in that service, so it needs to be using it as the network host.
When trying to set up this service without any of the wireguard/another service stuff, I got it working (locally), but when trying to use `network_mode`, since I can't use networks to set a static IP in the minecraft server, it doesn't seem to connect since the minecraft instance has `networks` and is not using `network_mode`, so it doesn't pass through to my wireguard container
this is a stripped down example of what I'm using in my compose:

```yaml
networks:
  wgnet:
    external: false
    driver: bridge
    ipam:
      config:
        - subnet: 10.130.157.0/24

services:
# VPN RELATED STUFF START
  wireguard:
    image: linuxserver/wireguard
    container_name: wireguard
    networks:
      wgnet:
        ipv4_address: 10.130.157.2
    cap_add:
      - NET_ADMIN
    volumes:
      - ${DOCKERCONFDIR}/wireguard:/config
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
      - TZ=${TZ}
    ports:
      - 51820:51820/udp
      - 25566:25566
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped

  lazymc:
    container_name: lazymc
    image: ghcr.io/joesturge/lazymc-docker-proxy:latest
    restart: unless-stopped
    volumes:
      - ${DOCKERCONFDIR}/mc-2:/server:ro
      - /var/run/docker.sock:/var/run/docker.sock:ro
    network_mode: "service:wireguard"

  mc-2:
    container_name: mc-2
    image: itzg/minecraft-server
    networks:
      wgnet:
        ipv4_address: 10.130.157.3
    labels:
      - lazymc.enabled=true
      - lazymc.group=mc
      - lazymc.server.address=mc-2:25566
    tty: true
    stdin_open: true
    environment:
      - EULA=TRUE
      - SERVER_NAME="TEST"
      - SERVER_PORT=25566
    volumes:
      - ${DOCKERCONFDIR}/mc-2:/data
    restart: no
```
This is not a problem related to this image, and more like a docker network thing, but I've been banging my head a few hours already and couldn't really get it to work, any help is appreciated.

If this is not the correct place to open this question, please let me know where and if needed close this issue too, thanks!

**Comment by joesturge** [comment]
Hi thanks for the question, I think it should work if you also give `network_mode: service:wireguard` to the Minecraft server and not use a static IP network at all. Then for server address label use: `lazymc.server.address=127.0.0.1:25566`

By doing this both lazymc and your server are both on the same host as wire guard so basically act like a server running in localhost.

Doing this is actually good, as the IP will always be the same still

**Comment by zhyph** [comment]
Thanks for the answer joe,

It kinda worked at first, resulting in a normal behavior, but after I refreshed my MC multiplayer page a few times, the following error happened, any ideas on how to debug this?
```bash
 INFO  lazymc-docker-proxy::entrypoint > Ensuring all server containers are stopped...
 WARN  lazymc-docker-proxy::docker     > **************************************************************************************************************************
 WARN  lazymc-docker-proxy::docker     > WARNING: You should use IPAM to assign a static IP address to your server container otherwise performance may be degraded.
 WARN  lazymc-docker-proxy::docker     >     see: https://github.com/joesturge/lazymc-docker-proxy?tab=readme-ov-file#usage
 WARN  lazymc-docker-proxy::docker     > **************************************************************************************************************************
 INFO  lazymc-docker-proxy::entrypoint > Starting lazymc process for group: mc...
 INFO  lazymc-docker-proxy::health     > Application is healthy.
 INFO  mc::lazymc                      > Proxying public 0.0.0.0:25565 to server 127.0.0.1:25565
 INFO  mc::lazymc::monitor             > Server is now online
 WARN  mc::lazymc                      > Failed to proxy: Address not available (os error 99)
 WARN  mc::lazymc                      > Failed to proxy: Address not available (os error 99)
 WARN  mc::lazymc                      > Failed to proxy: Connection reset by peer (os error 104)
 WARN  mc::lazymc                      > Failed to proxy: Connection reset by peer (os error 104)
 INFO  mc::lazymc::monitor             > Server is now sleeping
 INFO  mc::lazymc::monitor             > Server is now online
 WARN  mc::lazymc                      > Failed to proxy: Broken pipe (os error 32)
```

**Comment by joesturge** [comment]
Hi, when using `network_mode: service:whatever` you have to be careful not to use the same port between all containers on the same service network, as the containers act as if they are on the same host. I notice that the logs you sent indicate `proxying 0.0.0.0:25565 to 127.0.0.1:25565`. Assuming that lazymc is listening on port 25565, make sure that the Minecraft container is running on a different port.

TLDR; using service network mode makes containers essentially share hosts, so don't use the same ports on containers which share a service network.

**Comment by zhyph** [comment]
In this occasion, I have two servers running, one for "production" and one that I do tests with.
One has the default port on 25565, and the other one that I do tests with where I modified the server_port to 25566 (via env)

**Comment by joesturge** [comment]
Ah okay, so the lazymc container runs a TCP proxy for every server. So let's assume the following containers are all on the wire guard service network:
- lazymc:25565+25566
- production:25565
- test:25566

The issue is that lazymc proxy needs the 25565 port to run the TCP proxy for the prod server, so the prod server cannot run the server on port 25565. It should use an unused port (because service mode causes containers to run in same host, can't have two servers with single port). It's the same issue for the test server too.

I recommend, given your setup, to use the following labels on production server:
- lazymc.port=25565
- lazymc.server.address=127.0.0.1:25575 (use PORT=25575 on MC container)

And on the test server:
- lazymc.port=25566
- lazymc.server.address=127.0.0.1:25576 (use PORT=25576 on MC container)

This ensures that you do no reuse the same port more than once.

This is only an issue as lazymc, and the two servers all use network-mode:wireguard. So they cannot use the same port

Hope this helps

**Comment by zhyph** [comment]
Ohhhh now I see it, that definitely helps.

Thanks for the clarification, I'll give it a shot again with this new setup, will update in a bit.

**Comment by zhyph** [comment]
Hell yeah, this worked!
I just noticed that my minecraft server needs to already be running before actually starting lazymc, otherwise the proxy fails (I think so?), no apparent error logs so not sure, but good enough for me, I didn't test but using `depends_on` should probably do the trick, thank you so much!

**Comment by joesturge** [comment]
No worries, glad I could help 👍

Could you please share an example docker compose? Just so I can see what worked? And so that I can look into the behaviour you discussed

Cheers

**Comment by zhyph** [comment]
Hey Joe, sorry about the wait, here's the yml:
```yaml
networks:
  wgnet:
    external: false
    driver: bridge
    ipam:
      config:
        - subnet: 10.130.157.0/24

services:
# VPN RELATED STUFF START
  wireguard:
    image: linuxserver/wireguard
    container_name: wireguard
    networks:
      wgnet:
        ipv4_address: 10.130.157.2
    cap_add:
      - NET_ADMIN
    volumes:
      - ${DOCKERCONFDIR}/wireguard:/config
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
      - TZ=${TZ}
    ports:
      - 51820:51820/udp
      - 8080:8080
      - 25565:25565
      - 24454:24454/udp
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped

  lazymc:
    container_name: lazymc
    image: ghcr.io/joesturge/lazymc-docker-proxy:latest
    restart: unless-stopped
    volumes:
      - ${DOCKERCONFDIR}/mc:/server:ro
      - /var/run/docker.sock:/var/run/docker.sock:ro
    network_mode: "service:wireguard"

  mc:
    container_name: mc
    image: itzg/minecraft-server
    labels:
      - lazymc.port=25565
      - lazymc.enabled=true
      - lazymc.group=mc
      - lazymc.server.address=127.0.0.1:25566
    tty: true
    stdin_open: true
    environment:
      - EULA=TRUE
      - ONLINE_MODE=FALSE
      - INIT_MEMORY=1G
      - MAX_MEMORY=4G
      - TYPE=FABRIC
      - MODRINTH_ALLOWED_VERSION_TYPE=release
      - MODRINTH_DOWNLOAD_DEPENDENCIES=required
      - MODRINTH_PROJECTS=simple-voice-chat
      - TZ=${TZ}
      - SERVER_PORT=25566 # I've opted for using 25566 for the primary server, 25575 is the default RCON port
    volumes:
      - ${DOCKERCONFDIR}/mc:/data #DOCKERCONFDIR is the same as /docker/apppdata
    restart: no
    network_mode: "service:wireguard"
```

And the thing I said earlier, about needing to have minecraft start first and then start lazymc still holds, if start both of them at the same time lazymc will try to sleep the container immediately, which causes the server to never actually open (not sure how true this is, just a guess)

**Comment by zhyph** [comment]
Hey Joe, just updating, everything is working just fine, it was a mistake on my part, I misunderstood how the proxy actually works and thought something weird was happening
